### PR TITLE
Padding

### PR DESCRIPTION
**Pull request by yawn**
PR for https://github.com/coredns/coredns/issues/540

I've dropped the pinset support for the time being and will revisit it later.

Ready for review.

### CODE REVIEW

**Comment by johnbelamaric** [comment]
Convention is to put WIP (work in progress) in front if it's not ready for review.

**Comment by miekg** [comment]
We should still review this, no?

**Comment by yawn** [comment]
Yep :-)
Miek Gieben <notifications@github.com> schrieb am Do. 1. Juni 2017 um 19:13:

File: middleware/proxy/README.md [hunk 1]
```diff
@@ -26,7 +26,11 @@ proxy FROM TO... {
     health_check PATH:PORT [DURATION]
     except IGNORED_NAMES...
     spray
-    protocol [dns [force_tcp]|https_google [bootstrap ADDRESS...]|grpc [insecure|CA-PEM|KEY-PEM CERT-PEM|KEY-PEM CERT-PEM CA-PEM]]
+    protocol [dns [force_tcp]|grpc [insecure|CA-PEM|KEY-PEM CERT-PEM|KEY-PEM CERT-PEM CA-PEM]]
+    protocol https_google [bootstrap ADDRESS...] {
+      padding
+      pinset 0uXWIwuNdaaFSx7xJySymocwKDFJeEVzlouiy6MZATA=
```
Comment (miekg) [inline middleware/proxy/README.md:32]: pinset seems not to be used and even if we need it (I don't know what it is), it should be in another PR.

File: middleware/proxy/README.md [hunk 2]
```diff
@@ -54,6 +58,10 @@ proxy FROM TO... {
   the [DnsService](https://github.com/coredns/coredns/pb/dns.proto).
   An out-of-tree middleware that implements the server side of this can be found at
   [here](https://github.com/infobloxopen/coredns-grpc).
+  * `padding` sends random padding along with the request - the padding has a
+    fixed length of 255 minus the length of the name
```
Comment (miekg) [inline middleware/proxy/README.md:62]: drop "fixed" here and end the sentence with a dot.

File: middleware/proxy/padding.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	buffers sync.Pool
+}
+
+// NewPadding instantiates a new padding up to the given size (e.g. 256)
+func NewPadding(size uint16) *Padding {
+
+	// calculate the buffer we for a base64 representation that is "size" long
+	s := int(3.0 / 4.0 * float64(size))
+
+	return &Padding{
```
Comment (miekg) [inline middleware/proxy/padding.go:27]: Why all this? You just want to base64 encode some random string? I.e. `sync.Pool`?

File: middleware/proxy/padding.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// Generate produces an appropriatep adding string for the given name
+func (p *Padding) Generate(name string) (string, error) {
+
+	const empty = ""
+
+	padding := p.size - len(name)
+
+	if padding < 1 {
```
Comment (miekg) [inline middleware/proxy/padding.go:46]: Why this? According to https://developers.google.com/speed/public-dns/docs/dns-over-https padding makes the request the same size. So even if the name is 255 chars long we can still have padding if we make each message (say) 400 bytes.

Comment (miekg) [reply middleware/proxy/padding.go:46]: Think it makes sense to pad everything to 300 bytes no matter what.

File: middleware/proxy/padding.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		return empty, nil
+	}
+
+	buf := p.buffers.Get().(*bytes.Buffer)
```
Comment (miekg) [inline middleware/proxy/padding.go:50]: Ah, here is where you use the buffers... Is there a possibility to do this without all this buffer handling? Seems like overkill (but I could be wrong).

File: middleware/proxy/padding.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		buffers: sync.Pool{
+			New: func() interface{} {
+				return bytes.NewBuffer(make([]byte, s))
+			},
+		},
+	}
+
+}
+
+// Generate produces an appropriatep adding string for the given name
```
Comment (miekg) [inline middleware/proxy/padding.go:39]: needs closing dot and has typo in it.

**Review by miekg — CHANGES_REQUESTED** [review]
Thanks for the PR, look useful.

**Comment by miekg** [comment]
are you still working on this?

**Comment by miekg** [comment]
closing this, please re-open if you want to start working on it again.

**Comment by yawn** [comment]
Complete lack of time currently, sorry

**Pull request closed without merging**